src/category_theory/flat_functors.lean

```diff
@@ -0,0 +1,266 @@
+variables {J : Type v₁} [small_category J]
+variables {K : J ⥤ C} (F : C ⥤ D) (c : cone K)
+
+/--
+Given a cone `c : cone K` and a map `f : X ⟶ c.X`, we can construct a cone of structured
+arrows over `X` with `f` as the cone point. This is the underlying diagram.
+-/
+@[simps]
+def to_diagram : J ⥤ structured_arrow c.X K :=
+{ obj := λ j, structured_arrow.mk (c.π.app j),
+  map := λ j k g, structured_arrow.hom_mk g (by simpa),
+  map_id' := λ X, by simpa,
+  map_comp' := λ X Y Z g h, by { ext, simp } }
+
+/-- Given a diagram of `structured_arrow X F`s, we may obtain a cone with cone point `X`. -/
+@[simps]
+def diagram_to_cone {X : D} (G : J ⥤ structured_arrow X F) : cone (G ⋙ proj X F ⋙ F) := {
+  X := X, π := { app := λ j, (G.obj j).hom } }
+
+/--
+Given a cone `c : cone K` and a map `f : X ⟶ F.obj c.X`, we can construct a cone of structured
+arrows over `X` with `f` as the cone point.
+-/
+@[simps]
+def to_cone {X : D} (f : X ⟶ F.obj c.X) :
+  cone (to_diagram (F.map_cone c) ⋙ map f ⋙ pre _ K F) :=
+{ X := mk f, π := { app := λ j, hom_mk (c.π.app j) rfl,
+                    naturality' := λ j k g, by { ext, dsimp, simp } } }
+
+lemma eq_to_hom_right {A : Type*} [category A] {B : Type*} [category B] {T : Type*} [category T]
+  {L : A ⥤ T} {R : B ⥤ T} (X Y : comma L R) (H : X = Y) :
+  comma_morphism.right (eq_to_hom H) = eq_to_hom (by { cases H, refl }) := by { cases H, refl }
+
+local attribute[simp] eq_to_hom_right
+
+/--
+If a cone `s₁` factors through another cone `s₂`, then the two constructed diagrams are actually
+the same.
+-/
+lemma to_diagram_comp_map (s₁ s₂ : cone K)
+  (f : s₁.X ⟶ s₂.X) (H : ∀ (j : J), f ≫ s₂.π.app j = s₁.π.app j) :
+    to_diagram s₂ ⋙ structured_arrow.map f = to_diagram s₁ := by { apply functor.ext, tidy, }
+
+end structured_arrow_cone
+
+section representably_flat
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₂} D]
+
+/--
+A functor `F : C ⥤ D` is representably-flat functor if the comma category `(X/F)`
+is cofiltered for each `X : C`.
+-/
+class representably_flat (F : C ⥤ D) : Prop :=
+(cofiltered : ∀ (X : D), is_cofiltered (structured_arrow X F))
+
+instance functor.flat_cofiltered (F : C ⥤ D) [representably_flat F] (X : D) :
+ is_cofiltered (structured_arrow X F) := representably_flat.cofiltered X
+
+end representably_flat
+
+section has_limit
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₁} D]
+
+@[priority 100]
+instance cofiltered_of_has_finite_limit [has_finite_limits C] : is_cofiltered C :=
+{ cocone_objs := λ A B, ⟨limits.prod A B, limits.prod.fst, limits.prod.snd, trivial⟩,
+  cocone_maps :=  λ A B f g, ⟨equalizer f g, equalizer.ι f g, equalizer.condition f g⟩,
+  nonempty := ⟨⊤_ C⟩ }
+
+lemma flat_of_preserves_finite_limit [has_limits C] (F : C ⥤ D)
+  (H : ∀ (J : Type v₁) [h : small_category J] [h' : @fin_category J h],
+    @preserves_limits_of_shape _ _ _ _ J h F) : representably_flat F := ⟨λ X,
```

Comment:
We really should have an abbreviation/def for "preserves finite limits"...

Reply:
I've added it as a typeclass for now.